app/utils/JourneyChecker.scala

```diff
@@ -32,23 +33,48 @@ trait JourneyChecker extends IncomeSourcesUtils {
   implicit val ec: ExecutionContext
 
   val sessionService: SessionService
-  def withIncomeSourcesFSWithSessionCheck(journeyType: JourneyType)(codeBlock: => Future[Result])(implicit user: MtdItUser[_], hc: HeaderCarrier): Future[Result] = {
+
+  def withCustomSession(journeyType: JourneyType)(codeBlock: Option[UIJourneySessionData] => Future[Result])
+                       (implicit user: MtdItUser[_], hc: HeaderCarrier): Future[Result] = {
     withIncomeSourcesFS {
-      journeyChecker(journeyType).flatMap {
-        case true => user.userType match {
-          case Some(Agent) => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.showAgent(journeyType.businessType)))
-          case _ => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.show(journeyType.businessType)))
-        }
-        case false => codeBlock
+      sessionService.getMongo(journeyType.toString).flatMap{
+        case Right(Some(sessionData)) =>
+          customSessionStarted(sessionData).flatMap { state =>
+            (state, user.userType) match {
+              case (true, Some(Agent)) =>
+                Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.showAgent(journeyType.businessType)))
+              case (true, Some(Individual)) =>
+                Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.show(journeyType.businessType)))
+              case (_, _) => codeBlock(Some(sessionData))
+            }
+          }
+        case _ =>
+          codeBlock(None)
       }
     }
   }
 
-  private def journeyChecker(journeyType: JourneyType)(implicit hc: HeaderCarrier): Future[Boolean] = {
-    sessionService.getMongoKeyTyped[Boolean](AddIncomeSourceData.hasBeenAddedField, journeyType).flatMap {
-      case Right(Some(true)) => Future(true)
+  private def customSessionStarted(sessionData: UIJourneySessionData): Future[Boolean] = {
+    sessionData.addIncomeSourceData match {
+      case Some(value) if value.hasBeenAdded.contains(true) => Future(true)
```

Comment:
There is no need to directly create a future like this unless running a long-running async computation. This can lead to thread blocking and performance issues, see:
https://blog.colinbreck.com/future-successful-blocking-your-future-success/